core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/bi/BavetIfExistsBiConstraintStream.java

```diff
@@ -91,11 +89,11 @@ public <Score_ extends Score<Score_>> void buildNode(NodeBuildHelper<Score_> bui
         IfExistsBiWithUniNode<A, B, C> node = new IfExistsBiWithUniNode<>(shouldExist,
                 JoinerUtils.combineLeftMappings(joiner), JoinerUtils.combineRightMappings(joiner),
                 inputStoreIndexA, inputStoreIndexB,
-                insert, retract,
+                buildHelper.getAggregatedTupleLifecycle(childStreamList),
                 indexerAB, indexerC, filtering);
         buildHelper.addNode(node);
-        buildHelper.putInsertUpdateRetract(this, node::insertLeft, node::updateLeft, node::retractLeft);
-        buildHelper.putInsertUpdateRetract(parentBridgeC, node::insertRight, node::updateRight, node::retractRight);
+        buildHelper.putInsertUpdateRetract(this, TupleLifecycle.ofLeft(node));
+        buildHelper.putInsertUpdateRetract(parentBridgeC, TupleLifecycle.ofRight(node));
```

Comment:
And this is the case where we add an extra indirection. Joins and ifExists can not implement TupleLifecycle, because they have different left and right entry points, so we need to do this.

With Gizmo, we could probably generate this away. But benchmarking does not show this as a problem.